File: data/src/main/java/com/clloret/days/data/local/entities/DbEvent.java
package com.clloret.days.data.local.entities;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;
import com.clloret.days.data.local.entities.converters.TimeUnitConverter;
import com.clloret.days.domain.entities.Event.TimeUnit;
import java.util.Date;

@Entity(tableName = "events")
public class DbEvent {

  private static final String[] EMPTY_ARRAY = new String[0];

  @PrimaryKey
  @ColumnInfo(name = "id")
  @NonNull
  private String id;

  @ColumnInfo(name = "name")
  private String name;

  @ColumnInfo(name = "description")
  private String description;

  @ColumnInfo(name = "date")
  private Date date;

  @ColumnInfo(name = "tag_id")
  private String[] tags = EMPTY_ARRAY;

  @ColumnInfo(name = "favorite")
  private boolean favorite;

  @ColumnInfo(name = "reminder")
  @Nullable
  private Integer reminder;

  @ColumnInfo(name = "reminder_unit")
  @TypeConverters(TimeUnitConverter.class)
  private TimeUnit reminderUnit;

  @ColumnInfo(name = "time_lapse")
  private int timeLapse;

  @ColumnInfo(name = "time_lapse_unit")
  @TypeConverters(TimeUnitConverter.class)
  private TimeUnit timeLapseUnit;

  public DbEvent(@NonNull String id, String name, String description, Date date, boolean favorite,
      @Nullable Integer reminder, TimeUnit reminderUnit, int timeLapse, TimeUnit timeLapseUnit) {

    this.id = id;
    this.name = name;
    this.description = description;
    this.date = date;
    this.favorite = favorite;
    this.reminder = reminder;
    this.reminderUnit = reminderUnit;
    this.timeLapse = timeLapse;
    this.timeLapseUnit = timeLapseUnit;
  }

  @Ignore
  public DbEvent(@NonNull String id) {

    this.id = id;
  }

  @NonNull
  public String getId() {

    return id;
  }

  public void setId(@NonNull String id) {

    this.id = id;
  }

  public String getName() {

    return name;
  }

  public void setName(String name) {

    this.name = name;
  }

  public Date getDate() {

    return (Date) date.clone();
  }

  public void setDate(Date date) {

    this.date = (Date) date.clone();
  }

  public boolean isFavorite() {

    return favorite;
  }

  public void setFavorite(boolean favorite) {

    this.favorite = favorite;
  }

  public String[] getTags() {

    return tags;
  }

  public void setTags(String[] tags) {

    this.tags = tags;
  }

  public String getDescription() {

    return description;
  }

  public void setDescription(String description) {

    this.description = description;
  }

  public Integer getReminder() {

    return reminder;
  }

  public void setReminder(Integer reminder) {

    this.reminder = reminder;
  }

  public TimeUnit getReminderUnit() {

    return reminderUnit;
  }

  public void setReminderUnit(@NonNull TimeUnit reminderUnit) {

    this.reminderUnit = reminderUnit;
  }

  public int getTimeLapse() {

    return timeLapse;
  }

  public void setTimeLapse(int timeLapse) {

    this.timeLapse = timeLapse;
  }

  public TimeUnit getTimeLapseUnit() {

    return timeLapseUnit;
  }

  public void setTimeLapseUnit(TimeUnit timeLapseUnit) {

    this.timeLapseUnit = timeLapseUnit;
  }
}

File: data/src/main/java/com/clloret/days/data/local/entities/DbEvent.kt
package com.clloret.days.data.local.entities

import androidx.room.*
import com.clloret.days.data.local.entities.converters.TimeUnitConverter
import com.clloret.days.domain.entities.Event
import java.util.*

@Entity(tableName = "events")
class DbEvent {
  @PrimaryKey
  @ColumnInfo(name = "id")
  var id: String

  @ColumnInfo(name = "name", defaultValue = "")
  var name: String = ""

  @ColumnInfo(name = "description")
  var description: String? = null

  @ColumnInfo(name = "date")
  var date: Date = Date()

  @ColumnInfo(name = "tag_id")
  var tags: Array<String> = EMPTY_ARRAY

  @ColumnInfo(name = "favorite")
  var favorite = false

  @ColumnInfo(name = "reminder")
  var reminder: Int? = null

  @ColumnInfo(name = "reminder_unit", defaultValue = "Day")
  @TypeConverters(TimeUnitConverter::class)
  var reminderUnit: Event.TimeUnit = Event.TimeUnit.DAY

  @ColumnInfo(name = "time_lapse")
  var timeLapse = 0

  @ColumnInfo(name = "time_lapse_unit", defaultValue = "Day")
  @TypeConverters(TimeUnitConverter::class)
  var timeLapseUnit: Event.TimeUnit = Event.TimeUnit.DAY

  @ColumnInfo(name = "progress_date")
  var progressDate: Date? = null

  @JvmOverloads
  constructor(id: String,
              name: String,
              description: String?,
              date: Date,
              tags: Array<String> = EMPTY_ARRAY,
              favorite: Boolean,
              reminder: Int?,
              reminderUnit: Event.TimeUnit,
              timeLapse: Int,
              timeLapseUnit: Event.TimeUnit,
              progressDate: Date? = null) {
    this.id = id
    this.name = name
    this.description = description
    this.date = date
    this.tags = tags
    this.favorite = favorite
    this.reminder = reminder
    this.reminderUnit = reminderUnit
    this.timeLapse = timeLapse
    this.timeLapseUnit = timeLapseUnit
    this.progressDate = progressDate
  }

  @Ignore
  constructor(id: String) {
    this.id = id
  }

  companion object {
    private val EMPTY_ARRAY = emptyArray<String>()
  }
}